39 emailer

laurenbrissette opened the pull request:
Creating notifications -> emails based on user attributes or submission status.

Customized mails and notifications are created in database as logs.  Further testing needed to ensure the existing emailer is compatible with this new form of creating notifications (added a new notification type and email template type).

Changes:
- added ability to cc emails, and edit ccs individually
- added a reply email field, where replies will be directed -> otherwise, uses emailer's current response email taken from env
- students, instructors, and graders are filtered to choose which sections to include.  if no section, they are included 
- previews show each individual email that will be sent rather than a group 
- added template variables, which are resolved in the previews so instructor can preview full contents before send 
- history section at the bottom of the page, which blocks emails in bursts of 5 minutes.  shows 10 initially, with show more button to expand

<img width="1490" alt="Screenshot 2025-06-19 at 5 25 21 PM" src="https://github.com/user-attachments/assets/260fe327-dc76-4d93-ae0c-ec08cbdda578" />
<img width="1512" alt="Screenshot 2025-06-19 at 5 25 35 PM" src="https://github.com/user-attachments/assets/0e3b81f3-fb50-4bfe-88ae-699b2411ddff" />
<img width="1512" alt="Screenshot 2025-06-19 at 5 26 55 PM" src="https://github.com/user-attachments/assets/b11b0914-6470-4774-a7f7-744402a10c9a" />

laurenbrissette committed:
database support + basic display for email lists

laurenbrissette committed:
outstanding changes

laurenbrissette committed:
move from auto maintain -> fetch on request

laurenbrissette committed:
missing import

laurenbrissette committed:
new form format

laurenbrissette committed:
stage emails

laurenbrissette committed:
start to gather emails on send

laurenbrissette committed:
more filtering to find emails to send to

laurenbrissette committed:
bug fixes + make subject and body editable in staging

laurenbrissette committed:
drafting notification trigger on emails

laurenbrissette committed:
Merge branch 'staging' of https://github.com/pawtograder/platform into 39-emailer

laurenbrissette committed:
id undef. handling + split up finding ids

laurenbrissette committed:
email template + notification preview

laurenbrissette committed:
remove cc

laurenbrissette committed:
lint

jon-bell commented on app/course/[course_id]/manage/course/emails/previewAndSend.tsx:
```diff
@@ -0,0 +1,317 @@
+                  </Text>
+                </Flex>
+                <Flex flexDir={"column"} fontSize="sm">
+                  <Text>To: {emailToAudienceText(email)}</Text>
```
For tags, this throws browser error:

```
hook.js:608 In HTML, <div> cannot be a descendant of <p>.
This will cause a hydration error.
```

I think that it would be fine for this to be a `<Box>`, which would solve this error.

jon-bell reviewed (changes requested):
Very nice draft! Here are some high-level comments:

- I think that it would be useful to have a CC option when sending the email (can just enter some email addresses, and they will be CCed on the message, with a checkbox for "CC self"). Related: a "reply-to" option.
- I think that it will be very useful to be able to refer to some template variables in the subject and body of the email. For example, when sending to students who have/haven't submitted an assignment, template variables might include assignment name, the student's due date, a link to the assignment (or to a student's most recent submission) on Pawtograder, the student's group name (if any). 
- When selecting groups of users to send to directly (e.g. by selecting "Students"), it would be great to also be able to filter-down by section
- To remove any ambiguity into what emails will actually be sent, I think that the preview feature should show *all* of the templated emails. For example, if personalized emails will go to 10 students, then there would be 10 previews, 1 with each of the student's email addresses (and personalized information). The `email_recipients` table might be a good place to store the personalized subject and body.
- It would be very useful to be able to see a history of messages sent. The history should be grouped by the mail "batch" (in the example above, there's 1 batch of emails sent to 10 people, so the batch title might be "Message sent to 10 students on Jun 17 at 6:31AM", and then once its selected, it would show all 10 emails).

laurenbrissette committed:
pagination, preview by individual email + customization, template variables

laurenbrissette committed:
drafting history, foreign keys, student section filter

laurenbrissette committed:
history dashboard + start of cc support

laurenbrissette committed:
template variables

laurenbrissette committed:
move history + reply to support

laurenbrissette committed:
replies, cut redundant table, cleanup

laurenbrissette committed:
requested changes / cleanup

including
- mdeditor
- using pagesize
- callbacks in context
- memo for history cards
- organize history by batch
- add batches back to database
- template variables based on audience selection
- omit types
- redirect replies to email associated with their pawtograder account
- one call to db to create emails (multiple queued batches created individually to get ids)

also
- improved cc checkbox handling

laurenbrissette committed:
format

jon-bell committed:
Merge branch 'staging' into 39-emailer

jon-bell reviewed (approved):
LGTM, thanks for the very nice feature!
